### DISCUSSION TITLE
Unrecognizable lambda output in Cognito Pre authentication lambda trigger

### DISCUSSION THREAD

**Discussion by StephaneLeveugle** [brefphp/bref#1731] [Support]
**Description:**

Note: this issue is focused on pre authentication lambda trigger but I suppose it can happen everywhere AWS expects an object and receives an array

It is currently impossible to use the [Pre authentication lambda trigger](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-authentication.html) out of the box

It seems related to this issue https://github.com/brefphp/bref/issues/710, even though the event was a different one the logic behind it is the same

After some investigation the root cause seems to be PHP's array and how it behaves with json_encode

This is the payload received, which we should return as is if we accept the authentication

```json
{
    "request": {
        "userAttributes": {
            "string": "string",
            . . .
        },
        "validationData": {}
    },
    "response": {}
}
```

After converting the json to a PHP array and reencoding it to JSON to send to AWS the output becomes

```json
{
    "request": {
        "userAttributes": {
            "string": "string",
            . . .
        },
        "validationData": []
    },
    "response": []
}
```

which AWS considers an invalid lambda output with good reason

**How to reproduce:**

A basic lambda that just returns the event like this should allow every authentication

```php
<?php

declare(strict_types=1);
 
use Bref\Context\Context;
 
require __DIR__ . '/vendor/autoload.php';
 
return function (mixed $event, Context $context) {
    return $event;
};
```

However it fails for everyone with `Unrecognizable lambda output`

To fix it manually we can set the impacted fields to a Map or stdClass, which will result to the correct output when json encoded, example:

```php
<?php

declare(strict_types=1);
 
use Ds\Map;
use Bref\Context\Context;
 
require __DIR__ . '/vendor/autoload.php';
 
return function (mixed $event, Context $context) {
    $event['request']['validationData'] = new Map();
    $event['response'] = new Map();
    return $event;
};
```

Thank you :)

**Comment by mnapoli** [comment]
Hi! Unfortunately this is a problem that exists with PHP regardless of Lambda or not, Cognito, etc. You have to use `stdClass` instead of `[]` to return a JSON empty object. This is a workaround everyone has to do unfortunately, and I don't think Bref should fix this specific behavior here.

**Reply by StephaneLeveugle** [reply]
Hi,

I understand this isn't Bref's fault, still as a Bref user it would be nice to be able to use these events directly.
But I get it if you think this is out of scope.

What about a note in the documentation to warn end users of this unintended side effect of PHP's array?
As seen in https://github.com/brefphp/bref/issues/710, some people give up and end up using another language

**Reply by mnapoli** [reply]
A note in the docs would be great! Not sure where it would be best 🤔

**Reply by kkobold** [reply]
That note would have saved me a few hours today.   :D
